packages/core/src/lib/filter/index.ts

```diff
@@ -107,36 +110,87 @@ export class FilterCore extends BaseProtocol implements IBaseProtocolCore {
     );
 
     if (!res || !res.length) {
-      throw Error(
-        `No response received for request ${request.requestId}: ${res}`
-      );
+      return {
+        failure: {
+          error: ProtocolError.REMOTE_PEER_FAULT,
+          peerId: peer.id
+        },
+        success: null
+      };
     }
 
     const { statusCode, requestId, statusDesc } =
       FilterSubscribeResponse.decode(res[0].slice());
 
     if (statusCode < 200 || statusCode >= 300) {
-      throw new Error(
+      log.error(
         `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
       );
+      return {
+        failure: {
+          error: ProtocolError.REMOTE_PEER_REJECTED,
+          peerId: peer.id
+        },
+        success: null
+      };
     }
+
+    return {
+      failure: null,
+      success: peer.id
+    };
   }
 
   async unsubscribe(
     pubsubTopic: PubsubTopic,
     peer: Peer,
     contentTopics: ContentTopic[]
-  ): Promise<void> {
-    const stream = await this.getStream(peer);
+  ): Promise<CoreProtocolResult> {
+    let stream: Stream | undefined;
+    try {
+      stream = await this.getStream(peer);
+    } catch (error) {
+      log.error(
+        `Failed to get a stream for remote peer${peer.id.toString()}`,
+        error
+      );
+      return {
+        success: null,
+        failure: {
+          error: ProtocolError.REMOTE_PEER_FAULT,
+          peerId: peer.id
+        }
+      };
+    }
+
     const unsubscribeRequest = FilterSubscribeRpc.createUnsubscribeRequest(
       pubsubTopic,
       contentTopics
     );
 
-    await pipe([unsubscribeRequest.encode()], lp.encode, stream.sink);
+    try {
+      await pipe([unsubscribeRequest.encode()], lp.encode, stream.sink);
+    } catch (error) {
+      log.error("Failed to send unsubscribe request", error);
+      return {
+        success: null,
+        failure: {
+          error: ProtocolError.GENERIC_FAIL,
```

Comment:
shouldn't it be more concrete that `generic`?

Reply:
I think we should make errors as informative as possible (without creating a mess) as it is the only thing developer would get, so let's try not to use `GENERIC_FAIL` imho

Reply:
> shouldn't it be more concrete that `generic`?

no, this is also what we do during any lightpush request.

if `pipe` fails unexpectedly, we can't be sure of what went wrong.

Reply:
is there no response that we can read? I expect full node potentially send some message if it is on the side of it, but if failure is in pipe - then it can be generic (though in that case I still think there is a reason we can root cause)

can you repro unexpected fails of `pipe`?

Reply:
> can you repro unexpected fails of pipe?

no test vector as such for that


--

this try catch case actually handles error if error is **thrown** and there is no defined response from the pipe operation

if you follow the next few lines you will see we also handle errors based on bad values of `res`

Reply:
eg:
```
    try {
      res = await pipe(
        [query.encode()],
        lp.encode,
        stream,
        lp.decode,
        async (source) => await all(source)
      );
    } catch (err) {
      log.error("Failed to send waku light push request", err);
      return {
        success: null,
        failure: {
          error: ProtocolError.GENERIC_FAIL,
          peerId: peer.id
        }
      };
    }

    const bytes = new Uint8ArrayList();
    res.forEach((chunk) => {
      bytes.append(chunk);
    });

    let response: PushResponse | undefined;
    try {
      response = PushRpc.decode(bytes).response;
    } catch (err) {
      log.error("Failed to decode push reply", err);
      return {
        success: null,
        failure: {
          error: ProtocolError.DECODE_FAILED,
          peerId: peer.id
        }
      };
    }

    if (!response) {
      log.error("Remote peer fault: No response in PushRPC");
      return {
        success: null,
        failure: {
          error: ProtocolError.REMOTE_PEER_FAULT,
          peerId: peer.id
        }
      };
    }

    if (!response.isSuccess) {
      log.error("Remote peer rejected the message: ", response.info);
      return {
        success: null,
        failure: {
          error: ProtocolError.REMOTE_PEER_REJECTED,
          peerId: peer.id
        }
      };
    }
```

Reply:
I believe in that case `GENERIC_FAIL` on `pipe` means - failed to send a request, as it fails locally.

Reply:
> I believe in that case `GENERIC_FAIL` on `pipe` means - failed to send a request, as it fails locally.

hmm, could it fail while receiving a response for the remote peer? 
are we sure about that?

Reply:
dunno, I think we need to investigate it - if it bothers us